pubsub/google/cloud/pubsub_v1/publisher/batch/mp.py

```diff
@@ -0,0 +1,398 @@
+        self._client = client
+        self._manager = multiprocessing.Manager()
+
+        # Create a namespace that is owned by the client manager; this
+        # is necessary to be able to have these values be communicable between
+        # processes.
+        self._shared = self.manager.Namespace()
+        self._shared.futures = self.manager.list()
+        self._shared.messages = self.manager.list()
+        self._shared.message_ids = self.manager.dict()
+        self._shared.settings = settings
+        self._shared.status = self.Status.ACCEPTING_MESSAGES
+        self._shared.topic = topic
+
+        # This is purely internal tracking.
+        self._process = None
+        self._size = 0
+
+        # Continually monitor the thread until it is time to commit the
+        # batch, or the batch is explicitly committed.
+        if autocommit and self.settings.max_latency < float('inf'):
+            self._process = multiprocessing.Process(target=self.monitor)
+            self._process.start()
+
+    @property
+    def client(self):
+        """Return the client used to create this batch.
+
+        Returns:
+            ~.pubsub_v1.client.PublisherClient: A publisher client.
+        """
+        return self._client
+
+    @property
+    def manager(self):
+        """Return the client's manager.
+
+        Returns:
+            :class:`multiprocessing.Manager`: The manager responsible for
+                handling shared memory objects.
+        """
+        return self._manager
+
+    @property
+    def settings(self):
+        """Return the settings for this batch.
+
+        Returns:
+            ~.pubsub_v1.types.Batching: The settings for batch
+                publishing. These should be considered immutable once the batch
+                has been opened.
+        """
+        return self._shared.settings
+
+    @property
+    def status(self):
+        """Return the status of this batch.
+
+        Returns:
+            str: The status of this batch. All statuses are human-readable,
+                all-lowercase strings.
+        """
+        return self._shared.status
+
+    def commit(self):
+        """Asychronously publish all of the messages on the active branch.
+
+        This method may be safely called from the primary process.
+        """
+        process = multiprocessing.Process(self._commit)
+        process.start()
+
+    def _commit(self):
+        """Actually publish all of the messages on the active batch.
+
+        This moves the batch out from being the active batch to an in-flight
+        batch on the publisher, and then the batch is discarded upon
+        completion.
+        """
+        # Update the status.
+        self._shared.status = 'in-flight'
+
+        # Begin the request to publish these messages.
+        if len(self._shared.messages) == 0:
+            raise Exception('Empty queue')
+        response = self._client.api.publish(
```

Comment:
A scaling bottleneck to consider: our frontends will throttle individual gRPC channels connections to XX B/s; if you want to surpass this limit, you'll need to shard your traffic across multiple channels